File: common/src/main/java/org/sandboxpowered/silica/network/login/serverbound/EncryptionResponse.java
package org.sandboxpowered.silica.network.login.serverbound;

import org.sandboxpowered.silica.network.*;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.Key;

public class EncryptionResponse implements Packet {
    public byte[] sharedSecret;
    public byte[] verifyToken;

    private static Cipher setupCipher(int i, String string, Key key) throws Exception {
        Cipher cipher = Cipher.getInstance(string);
        cipher.init(i, key);
        return cipher;
    }

    private static byte[] cipherData(int i, Key key, byte[] bs) throws EncryptionException {
        try {
            return setupCipher(i, key.getAlgorithm(), key).doFinal(bs);
        } catch (Exception var4) {
            throw new EncryptionException(var4);
        }
    }

    public static byte[] decryptUsingKey(Key key, byte[] bs) throws EncryptionException {
        return cipherData(2, key, bs);
    }

    @Override
    public void read(PacketByteBuf buf) {
        sharedSecret = buf.readByteArray();
        verifyToken = buf.readByteArray();
    }

    @Override
    public void write(PacketByteBuf buf) {
        buf.writeByteArray(sharedSecret);
        buf.writeByteArray(verifyToken);
    }

    @Override
    public void handle(PacketHandler packetHandler, Connection connection) {
        connection.handleEncryptionResponse(this);
    }

    public SecretKey getSecretKey(Key key) throws EncryptionException {
        byte[] cs = decryptUsingKey(key, this.sharedSecret);

        try {
            return new SecretKeySpec(cs, "AES");
        } catch (Exception var4) {
            throw new EncryptionException(var4);
        }
    }

    public byte[] getVerificationToken(Key key) throws EncryptionException {
        return decryptUsingKey(key, this.verifyToken);
    }

    public Cipher getCipher(int i, Key key) throws EncryptionException {
        try {
            Cipher cipher = Cipher.getInstance("AES/CFB8/NoPadding");
            cipher.init(i, key, new IvParameterSpec(key.getEncoded()));
            return cipher;
        } catch (Exception var3) {
            throw new EncryptionException(var3);
        }
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/vanilla/network/login/serverbound/EncryptionResponse.kt
package org.sandboxpowered.silica.vanilla.network.login.serverbound

import org.sandboxpowered.silica.vanilla.network.*
import java.security.Key
import javax.crypto.Cipher
import javax.crypto.SecretKey
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

class EncryptionResponse(private var sharedSecret: ByteArray, private var verifyToken: ByteArray) : Packet {
    constructor(buf: PacketByteBuf) : this(buf.readByteArray(), buf.readByteArray())

    override fun write(buf: PacketByteBuf) {
        buf.writeByteArray(sharedSecret)
        buf.writeByteArray(verifyToken)
    }

    override fun handle(packetHandler: PacketHandler, connection: Connection) {
        connection.handleEncryptionResponse(this)
    }

    @Throws(EncryptionException::class)
    fun getSecretKey(key: Key): SecretKey {
        val cs = decryptUsingKey(key, sharedSecret)
        return try {
            SecretKeySpec(cs, "AES")
        } catch (var4: Exception) {
            throw EncryptionException(var4)
        }
    }

    fun getVerificationToken(key: Key): ByteArray {
        return decryptUsingKey(key, verifyToken)
    }

    fun getCipher(i: Int, key: Key): Cipher {
        return try {
            val cipher = Cipher.getInstance("AES/CFB8/NoPadding")
            cipher.init(i, key, IvParameterSpec(key.encoded))
            cipher
        } catch (var3: Exception) {
            throw EncryptionException(var3)
        }
    }

    private fun setupCipher(i: Int, string: String, key: Key): Cipher {
        val cipher = Cipher.getInstance(string)
        cipher.init(i, key)
        return cipher
    }

    private fun cipherData(i: Int, key: Key, bs: ByteArray): ByteArray {
        return try {
            setupCipher(i, key.algorithm, key).doFinal(bs)
        } catch (var4: Exception) {
            throw EncryptionException(var4)
        }
    }

    private fun decryptUsingKey(key: Key, bs: ByteArray): ByteArray {
        return cipherData(2, key, bs)
    }
}